### DISCUSSION TITLE
Image characterization preprocessor

### DISCUSSION THREAD

**Discussion by kba** [OCR-D/core#761] [Ideas]
As discussed in our Open Tech Call, there is interest in the community for a pre-processor that validates certain assumptions on the images and possibly offers solutions to alleviate them. This arose from a discussion around reliable calculation of DPI (cf. #676).

Maybe we can use this thread to gather requirements before moving forward with the implementation.

Functionality such a processor should have:

* Determine the pixel density in DPI for all images
    * Parameters:
        * `min-dpi`: Lower threshold to reject an image
        * `estimate-dpi`: Whether to use heuristics on image size and page size to estimate DPI
        * `scale`: Whether to scale images to fake a higher DPI
* Determine the hardware capabilities of the underlying system
    * Can this image be processed with the RAM or GPU memory available?

Feedback and proposals are welcome!

**Comment by bertsky** [comment]
IIRC we also wanted to validate image formats here, i.e. 
- whether the image containers/codecs themselves are permitted, 
- whether the image metadata are valid and consistent (e.g. PNG header vs EXIF info),
- whether the image dimensions are permitted for a certain workflow or processing node (as lower or upper limit),
- whether the image colours are allowed (e.g. alpha channels, CMYK and others)

The idea was to get a **fail-early-and-loud** mechanism, but as opt-in (by using this processor as first step in the workflow). Another idea might be to conveniently **repair**, if possible. (This aspect applies to DPI meta-data as well).

**Comment by bertsky** [comment]
BTW, DPI estimation is non-trivial and already implies some kind of line segmentation dependency (like ocrolib) IMO.

Strategically, we could approach this as a processor in core with some base functionality (without DPI estimation, or a very crude one), but hooks for easy overriding. Other modules (like ocrd_cis) could then subclass from the image characterization in core to provide their own.
